_posts/2019-07-04-cbo-introduction.md

```diff
@@ -0,0 +1,285 @@
+
+## Filter statistics
+
+![](/assets/blog/cbo-introduction/filter-statistics.png)
+
+When considering a filtering operation, a filter’s condition is analyzed and
+the following estimations are calculated:
+
+- what is the probability that data row will pass the filtering condition. From
+  this, expected number of rows after the filter is derived,
+- fraction of `NULL` values for columns involved in the filtering condition (for
+  most conditions, this will simply be 0%),
+- number of distinct values for columns involved in the filtering condition,
+- number of distinct values for columns that were not part of the filtering
+  condition, if their original number of distinct values was more than the
+expected number of data rows that pass the filter.
+
+For example, for a condition like `l.item = 106170` we can observe that:
+
+- no rows with `l.item` being `NULL` will meet the condition,
+- there will be only one distinct value of `l.item` (106170) after the
+  filtering operation,
+- on average, number of data rows expected to pass the filter will be equal to
+  `number_of_input_rows * fraction_of_non_nulls / distinct_values`. (This
+  assumes, of course, that users most often drill down in the data they really
+  have, which is quite a reasonable assumption and also safe to make).
+
+## Projection statistics
+
+![](/assets/blog/cbo-introduction/projection-statistics.png)
+
+Projections (`l.item – 1 AS iid`) are similar to filters, except that, of
+course, they do not impact the expected number of rows after the operation.
+
+For a projection, the following types of column statistics are calculated (if
+possible for given projection expression):
+
+- number of distinct values produced by the projection,
+- fraction of `NULL` values produced by the projection,
+- minimum/maximum value produced by the projection.
+
+Naturally, if `iid` is only returned to the user, then these statistics are not
+useful. However, if it’s later used in filter or join operation, these
+statistics are important to correctly estimate the number of rows that meet the
+filter condition or are returned from the join.
+
+# Conclusion
+
+Summing up, Presto’s Cost-Based Optimizer is conceptually a very simple thing.
+Alternative query plans are considered, the best plan is chosen and executed.
+Details are not so simple, though. Fortunately, to use
+[Presto](https://prestosql.io/), one doesn’t need to know all these details.
+Of course, anyone with a technical inclination that like to wander in database
+internals is invited to study [the Presto
+code](https://github.com/prestosql/presto)!
+
+Enabling Presto CBO is really simple:
+
+- set `optimizer.join-reordering-strategy=AUTOMATIC` and
+  `join-distribution-type=AUTOMATIC` in your `config.properties`,
+- [analyze](https://prestosql.io/docs/current/sql/analyze.html) your tables,
+- no, there is no third step. That’s it!
+
+Take Presto CBO for a spin today and let us know what is *your* Presto
+experience!
```

Comment:
Even though twitter link is in the footer, i think it would be valuable to add something directly here. (People don't see the footer, right?)